### DISCUSSION TITLE
[Part -1] Search as you type | ashish.one

### DISCUSSION THREAD

**Discussion by giscus** [ashishtiwari1993/ashish.one#17] [General]
# [Part -1] Search as you type | ashish.one

Introduction In this blog, we will try to understand how “Search as you type” works and Quickly setup one demo using some sample data. You must have seen various websites like eCommerce, food apps, etc. where you just start typing & simultaneously relevant options start displaying as suggestions and autocomplete. We will try to achieve somewhat the same feature. Search as you type Elasticsearch gives this specific mapping type which you can simply set to a specific field where you want to perform this kind of search.

https://ashish.one/blogs/search-as-you-type/

**Comment by ross-abramson** [comment]
What about match_phrase_prefix with four terms? Eg: using the same example above, match_phrase prefix("best jogging shoes f")-- will this work or fail since edge gram only goes up to three tokens?

What about match_phrase prefix("best jogging shoes for m")?

**Reply by ashishtiwari1993** [reply]
Hi @ross-abramson,

Sorry for delay in reply!!

So search with `best jogging shoes f` and `best jogging shoes for m` both will give you the output. Because `match_phrase_prefix` return the result which matching with the terms in same order as provided.

So in case of `best jogging shoes f` it will match with strict order till `best jogging shoes` after that it will match with top 50 terms which will start with `f` which is

```
  "f",
  "fo",
  "for",
  "for ",
  "for m",
  "for me",
  "for men",
  "for men ",
```

For example

search query - `jogging shoes f`

It will would match a value of `jogging shoes` or `best jogging shoes for` but not with `best jogging tshirt and shoes`

**Comment by mat-kop** [comment]
Hi Ashish, is there a way to have AND instead of OR for returned matches?
What I mean is that in the employee data gif up top, it would only return employees that match all partial/full strings in the query.

**Reply by ashishtiwari1993** [reply]
@max-kop have you tried `match_phrase_prefix` query? It is going to match in strict order.

**Comment by yousef8** [comment]
Hi, thanks for the great article, This article is what i was looking for, but there is a little adjustment i want to achieve in search behavior.

Can i search for substrings not at the beginning of the word in the middle or end, meaning you searched for `jog` and it worked because it's as the beginning for the word what if i typed `ggi` which is in the middle of the word `jogging` and i want to get all results with words containing the part `ggi` anywhere in the word, is this possible?

**Reply by ashishtiwari1993** [reply]
Hey @yousef8, Yes you can achieve this by using [n-gram tokenizer](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-ngram-tokenizer.html). But be careful while defining value for `min_gram` & `max_gram`. It may increase the [storage size](https://medium.com/@ashishstiwari/what-should-be-the-value-of-max-gram-and-min-gram-in-elasticsearch-f091404c9a14). You can better control if you know what type of search query you would like to cater.
